### ISSUE TITLE
Could embeddings be dynamically constructed with reserved weights to indicate end of word / word with comma after / etc to save tokens?

### ISSUE THREAD

**Issue by CodeExplode** [meta-llama/llama#1072]
I save a lot of time in training Stable Diffusion models by pre-finetuning embeddings to capture as much of the concepts as possible. However it would be far easier if some weights had specific meanings at various value intervals (person, animal, etc), leaving fewer trainable parameters, and making it easier to set the right place in the distribution. Presumably the model would have an easier time learning as well, with consistent meanings of many of the weights from the get-go, saving a ton of training time.

In terms of LLMs, it seems like tokenization could use this strategy too, populating some weights of the embedding manually.

e.g. A single weight could indicate end of word (0.0), followed by comma (0.05), followed by period (0.1), followed by semicolon (0.2), and save a ton of tokens when tokenizing text, especially code. Every symbol doesn't need to be there, e.g. opening and closing brackets could be left as standalone tokens.

Expanding on this, encoders with a bottleneck are quite good at compressing information down to just a few latent values, which could be non-trainable weights on these dynamic embeddings. 2 or 3 could be dedicated to encoding consonants to help the model understand how words sound, with a big focus on the loss of the last consonant when training the encoder, to help with the model being able to rhyme. These do not need to be encoded for every word during inference, and can be cached, unless the model encounters a new word.

Similarly, a few weights could even be given basic pictorial representations of concepts associated with a word. Stable Diffusion 1.x's VAE can encode 8x8 RGB pixels to 4 latent values fairly successfully, how much could be encoded using a very simple silhouette/flat colours example system? This way you're giving the model quite a lot of context to work with which humans get in the real world through life, e.g. the shape of an egg, and could maybe even test it through drawing ascii characters. This becomes difficult for homonyms and words which exist in multiple languages, but it doesn't need to be perfect, and the model can learn that terms like "The Rock" don't refer to a literal rock, as usual. Perhaps it could even request pictorial weights at a later stage in the model, once it has determined the meaning of the word. This can again be mostly cached.

Another weight could indicate that the word starts with a capital (to learn nouns), is all capitals (to learn acronyms), is single letters separated by period characters, etc. The model would presumably learn that the lack of an all-uppercase flag can still mean an acronym spelled in lower case, but would have an easier time of learning the concept from the get-go, seeing near identical embeddings for the same concept with one minor difference.
